### ISSUE TITLE
Out of memory when training onset frames model

### ISSUE THREAD

**Issue by falaktheoptimist** [tensorflow/magenta#1615]
Hi!
We have created a TFrecords dataset for the onset frames model using LMD dataset using a slightly modified onsets_frames_transcription_create_dataset_maps script. The generated tfrecord contains 2560 records and max size of a single record is 8 MB. (Entire tfrecord file is of 1.5 GB). We then tried training the model with batch size of 8, 6 and even down to 4. However, the training always crashes with resource exhaused OOM error. We tried training on Titan X / Xp both with 12 GB RAM. Not sure what is the issue or what bit consumes the so much memory.
Relevant line from log when training with batch size 4

> 2019-11-04 12:22:31.814147: W tensorflow/core/common_runtime/bfc_allocator.cc:239] Allocator (GPU_0_bfc) ran out of memory trying to allocate 1.80GiB with freed_by_count=0. The caller indicates that this is not a failure, but may mean that there could be performance gains if more memory were available.
> 2019-11-04 12:22:41.854500: W tensorflow/core/common_runtime/bfc_allocator.cc:419] Allocator (GPU_0_bfc) ran out of memory trying to allocate 405.96MiB (rounded to 425677824).  Current allocation summary follows.

> 2019-11-04 12:22:41.861425: I tensorflow/core/common_runtime/bfc_allocator.cc:923] total_region_allocated_bytes_: 11838898176 memory_limit_: 11838898176 available bytes: 0 curr_region_allocation_bytes_: 23677796352
> 2019-11-04 12:22:41.861443: I tensorflow/core/common_runtime/bfc_allocator.cc:929] Stats: 
> Limit:                 11838898176
> InUse:                 11419856128
> MaxInUse:              11420587776
> NumAllocs:                   25815
> MaxAllocSize:           3997948672
> 
> 2019-11-04 12:22:41.861516: W tensorflow/core/common_runtime/bfc_allocator.cc:424] *********************************************************************************__*****************
> 2019-11-04 12:22:41.861560: W tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at fused_batch_norm_op.cc:1108 : Resource exhausted: OOM when allocating tensor with shape[4,96,2431,114] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc

Also, common pattern observed when the training crashes is that the last 2 dimensions are always 2431,114. 
Thank you in advance.

**Comment by cghawthorne** [comment]
I'd try setting the batch size to 1 and maybe trying out different values for truncated_length_secs until you get something that works and then try scaling up from there.

Did you end up changing anything about the size of the pianorolls, like adding more supported notes or something? That could definitely affect things.

**Comment by falaktheoptimist** [comment]
No I didn't add any more supported notes. It's still at default value. Trying with batch size 1 and keeping truncated_length_secs same it ran for about 1300 steps before the same OOM error. I'm suspecting a gradual filling up of memory at each sample somewhere along the way because even with the big network size, 12 GB wouldn't get filled up this quickly. Will post updates here as I dig more through the code.
Just curious - what GPU and batch size was the original model trained on?

**Comment by falaktheoptimist** [comment]
No I didn't add any more supported notes. It's still at default value. Trying with batch size 1 and keeping truncated_length_secs same it ran for about 1300 steps before the same OOM error. I'm suspecting a gradual filling up of memory at each sample somewhere along the way because even with the big network size, 12 GB wouldn't get filled up this quickly. Will post updates here as I dig more through the code.
Just curious - what GPU and batch size was the original model trained on?

**Comment by falaktheoptimist** [comment]
No I didn't add any more supported notes. It's still at default value. Trying with batch size 1 and keeping truncated_length_secs same it ran for about 1300 steps before the same OOM error. I'm suspecting a gradual filling up of memory at each sample somewhere along the way because even with the big network size, 12 GB wouldn't get filled up this quickly. Will post updates here as I dig more through the code.
Just curious - what GPU and batch size was the original model trained on?

**Comment by cghawthorne** [comment]
Hmm, that does seem odd. We used a variety of training setups, and don't remember for sure what we used for that checkpoint, but something along the lines of a V100 GPU with batch size 8 (but we might have done data parallelism across several GPUs).

Maybe it's a TF bug? Could you try an older build of TF?

**Comment by falaktheoptimist** [comment]
I tried a different builds - I don't think that was the bug. In one case it was the custom ftrecord we had built for LMD data which was having some records with more than 20 seconds of wav data which was causing it. Different truncated_length_secs did the trick in that case and got it working with batch size 6. Thank you so much for the help!
Our training is going on, but is slow with single GPU for dataset of the size of maestro. Thanks again.
